parallel-sums.1
.TH PARALLEL-SUMS 1 "November 2019" "Jigit jigdo tools"
.SH NAME
parallel-sums \- Calculate checksums in parallel using multiple algorithms
.SH SYNOPSIS
.B parallel-sums
\-\-algo1 \fBALGO1FILE\fR
[\-\-algo2 \fBALGO2FILE\fR ... \-\-algoN \fBALGONFILE\fR]
file1 [file2 ... fileN
.PP
.B parallel-sums
\-\-help
.SH DESCRIPTION
.PP
parallel-sums is designed to make it faster and easier to create
checksum files using multiple algorithms. It will only read the data
from each input file once, spawning a new thread per checksum
algorithm requested and running calculations in parallel.
.PP
Specify the list of desired checksum algorithms along with the output
filename for each checksum type, then pass the list of files.
.PP
The output is intended to be 100% compatible with the file formats of
existing tools like md5sum, sha256sum etc.
.TP
\fB\-\-algo ALGOSUMS\fR
Checksum all input files using the algorithm "algo" and output the
checksums in the file ALGOSUMS.
.TP
\fB\-\-help\fR
Print usage information and list the checksum algorithms supported.
.SH "EXAMPLES"
.TP
\fBparallel-sums \-\-md5sum MD5SUMS \-\-sha256sum SHA256SUMS *.iso\fR
Read all files matching the pattern *.iso. In parallel, checksum each
file using the MD5 algorithm (writing to "MD5SUMS") and checksum each
file using the SHA256 algorithm (writing to "SHA256SUMS").
.SH "SEE ALSO"
\fBmd5sum\fP(1), \fBsha1sum\fP(1), \fBsha256sum\fP(1) and \fBsha512sum\fP(1).
.SH "COPYRIGHT"
Copyright 2017-2019 Steve McIntyre (steve@einval.com)
.PP
parallel-sums may be copied under the terms and conditions of version
2 of the GNU General Public License, as published by the Free Software
Foundation (Cambridge, MA, USA).
.SH "AUTHOR"
Written by Steve McIntyre (steve@einval.com)


ChangeLog
Version 1.22 (2019-11-21)

 * Add support for SHA256 checksums, using jigdo format v2
   + Changes to libjte to generate the new format as an option - still
     defaults to v1 for cmpatibility for now
   + libjte updated to major version 2
   + Add support for v2 in jigdump, jigit-mkimage etc. too
   + Add new jigsum-sha256 program, for the base64-like output with
     sha256 checksums
 * jigdump now also understands jigdo .iso.tmp files
 * Misc small cleanups

Version 1.21 (2019-02-04)

 * parallel-sums:
   + New utility for calculating checksums in parallel
 * Misc code cleanups
 * Misc man page cleanups
 * jigit-mkimage:
   + Print missing file names correctly

Version 1.20 (2014-08-18)

 * mkjigsnap:
   + Add support for tracking back-references, i.e. which jigdo did a
     particular file come from.
   * Add support for verifying checksums
 * jigsum:
   * Add support for verifying checksums, similar to md5sum etc.
 * Misc code cleanups:
   * warnings
   * sparse errors
   * Fix memset calls to zero the right amount of memory

Version 1.19 (2011-06-16)

 * Improved error-handling in mkimage and libjte.
 * mkjigsnap rewrite:
   + merge two old implementations
   + switch to perl: cleaner code, much faster snapshot generation
   + added -v (verbose) and -N (dry-run) options
 * Put AC_CONFIG_MACRO_DIR in a better place in the libjte configure
   script.
 * Remove -Werror from the Makefile - gcc is getting more and more
   pedantic with every release, and this may cause package builds to fail.

Version 1.18 (2011-04-30)

 * More dynamic memory handling in mkimage, instead of allocating
   fixed-size strings on the stack. Should now work on Hurd too.
 * Fix man page syntax errors.
 * s,config.h,../config.h, in libjte for easier use by the xorriso folks

Version 1.17 (2011-03-16)

 * Rename mkimage to jigit-mkimage, to avoid a naming clash.
 * Sanity-check jigdo files - require a valid JigsawDownload
   header by default.
 * Multiple fixes for how MD5 comparisons are dealt with.
 * Add libjte - jigdo generation code that used to be embedded in
   mkisofs/genisoimage. Now a shared library useable in xorriso etc.

Version 1.16 (2007-10-18)

 * Remove build-stamp from the release tarball, to make debian package
   builds work better
 * Fix wget options typo
 * Add proper working bzip2-compressed template support in jigdump and
   mkimage
 * In mkimage, don't stat every file mentioned in the md5 list until we
   need to.
 * Other small cleanups in jigdump and mkimage

Version 1.15 (2005-01-17)

 * Minor compilation fixes
 * Cope with missing files better
 * Minor jigdo file-handling cleanups
 * mkimage code refactored to make it easier to maintain
 * Cleaned up licensing - added explicit GPL v2 headers all over
 * Fix image reconstruction with jigit

Version 1.14 (2004-10-12)

 * Fixed generation of rsyncsums
 * Added -d option to mkjigsnap to specify DIRNAME
 * Minor prettiness tweaks to output UI of mkimage and jigit
 * 64-bit printf fixes

Version 1.13 (2004-09-05)

 * Improved the documentation, listing new options to mkimage
 * Fixed/updated the debian-cd patch

Version 1.12 (2004-09-05)

 * Improved the documentation, listing new options to mkimage
 * Fixed/updated the debian-cd patch

Version 1.11 (2004-09-05)

 * Added rsync sums into JTE output
 * Added jigit script
 * Added extra options to mkimage to support jigit

Version 1.10 (2004-09-01)

 * Removed the last md5 check; no longer needed
 * Added new -jigdo-force-md5 option; any files matching specified
   paths MUST match an MD5 supplied, or fail
 * Fixed a couple of typos.
 * Added initial work for bz2 support.
 * debian-cd diff now includes updated grab_md5 and jigdo_cleanup
   scripts that must be used.
 * mkisofs now leaves a blank [Servers] section in the .jigdo file -
   jigdo_cleanup does the work.

Version 1.9 (2004-08-30)

 * Added MD5 reverse lookups to pick up boot files; helps reduce
   CD/DVD #1 size substantially, BUT -md5-list option now mandatory.
 * Fixed sizing bug in HFS hybrid template code.

Version 1.8 (2004-08-27)

 * Tweaked the -exclude option to reduce template file sizes.

Version 1.7 (2004-08-19)

 * Fixed a couple of silly bugs introduced; don't assume template
   creation.
 * Added README files and man page sections for the new boot
   support.

Version 1.6 (2004-07-14)

 * Added code to check MD5 sums in mkisofs; should no longer need
   to run mirror_check any more. See the -md5-list option.

Version 1.5 (2004-07-05)

 * Minor bug-fixes and updates to mkimage
 * Added iso-image.pl wrapper script

Version 1.4 (2004-06-26)

 * Added mkimage
 * Moved utils and patch into a tarball

Version 1.3 (2004-06-18)

 * Fixed crash without jigdo options
 * Cleaned up mipsel boot code
 * Make it easier to specify the boot command line for hppa
 * Minor cosmetic .jigdo cleanup

Version 1.2 (2004-06-14)

 * Added boot support for alpha, hppa, mips, mipsel
 * Cosmetic changes to the .jigdo file

Version 1.1 (2004-06-10)

 * Removed .jte file code
 * Large cleanup to reduce patch size

Version 1.0 (2004-06-07)

 * Initial release
 * Produced .jigdo, .template and .jte files


parallel-sums.c
/*
 * parallel-sums
 *
 * Calculate and print various checksums in parallel
 *
 * Copyright (c) 2015 - 2019 Steve McIntyre <steve@einval.com>
 *
 * GPL v2 - see COPYING
 */

#include <stdio.h>
#include <errno.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include <getopt.h>
#include <pthread.h>
#include <sys/time.h>

static char *progname = "parallel-sums";

#define THREADED_CHECKSUMS
#include "libjte/checksum.h"

#ifndef MIN
#define MIN(x,y)        ( ((x) < (y)) ? (x) : (y))
#endif

struct state
{
    int desired;
    char *out_filename;
    FILE *out_stream;
};

/* Simple circular buffer implementation */
#define QSIZE 4
#define BUF_SIZE (1024*1024)
struct buf 
{
    unsigned long long bytes;
    unsigned char      d[BUF_SIZE];
};
#define NUM_BUF (QSIZE + 2)
struct buf data[NUM_BUF];

struct {
    pthread_mutex_t mutex;
    int read_idx; /* index in the data[]array for the reader to point at */
    int put_idx;
    int get_idx;
    int num_full;
    void *data[QSIZE];
    pthread_t reader;
    pthread_t processor;	
    pthread_cond_t notfull;
    pthread_cond_t notempty;
    FILE *infile;
} queue;

#define TIME long long
TIME total_read = 0;
TIME total_proc = 0;
unsigned long long bytes_read;
unsigned long long bytes_proc;
static struct state state[NUM_CHECKSUMS];
static struct option opts[NUM_CHECKSUMS + 2];
int checksums_mask = 0;

static void get_time(TIME *newtime)
{
    struct timeval tv;
    struct timezone tz;

    tz.tz_minuteswest = 0;
    tz.tz_dsttime = 0;    

    gettimeofday(&tv, &tz);
    *newtime = ((TIME)tv.tv_sec * 1000000) + tv.tv_usec;
}

void put_cb_data(void *data)
{
    pthread_mutex_lock(&queue.mutex);
    /* wait while the buffer is full */
    while (queue.num_full == QSIZE)
        pthread_cond_wait(&queue.notfull, &queue.mutex);
    queue.data[queue.put_idx] = data;
    queue.put_idx++;
    if (queue.put_idx >= QSIZE)
        queue.put_idx = 0;
    queue.num_full++;
    /* let a waiting consumer know there is data */
    pthread_cond_signal(&queue.notempty);
    pthread_mutex_unlock(&queue.mutex);
}

void *get_cb_data(void)
{
    void *data;

    pthread_mutex_lock(&queue.mutex);
    /* wait while there is nothing in the buffer */
    while (queue.num_full == 0)
        pthread_cond_wait(&queue.notempty, &queue.mutex);
    data = queue.data[queue.get_idx];
    queue.get_idx++;
    if (queue.get_idx >= QSIZE)
        queue.get_idx = 0;
    queue.num_full--;
    /* let a waiting producer know there is room */
    pthread_cond_signal(&queue.notfull);
    pthread_mutex_unlock(&queue.mutex);
    return data;
}

void *reader(void * junk)
{
    TIME start,stop,diff;
    while (1)
    {
        struct buf *in_ptr = &data[queue.read_idx];
        get_time(&start);
        in_ptr->bytes = fread(in_ptr->d, 1, BUF_SIZE, queue.infile);
        get_time(&stop);
        bytes_read += in_ptr->bytes;
        diff = stop - start;
	total_read += diff;
        put_cb_data(in_ptr);
        queue.read_idx++;
	if (queue.read_idx >= NUM_BUF)
            queue.read_idx = 0;

        if (in_ptr->bytes == 0)
            return NULL;
    }
}

void *processor(checksum_context_t *ctx)
{
    TIME start,stop,diff;
    while (1)
    {
        struct buf *ptr = get_cb_data();
        if (ptr && ptr->bytes)
        {
            get_time(&start);
            checksum_update(ctx, ptr->d, ptr->bytes);
            get_time(&stop);
            diff = stop - start;
            total_proc += diff;
            bytes_proc += ptr->bytes;
        }
        else
            return NULL;
    }
}

void usage()
{
    int i = 0;

    printf("%s\n\n", progname);
    printf("Usage: parallel-sums <--algo1 ALGO1FILE> [--algo2 ALGO2FILE ... -algo2N algoNfile] \\\n");
    printf("       <file1> [file2 ... fileN]\n\n");

    printf("Calculate checksums for a set of files (e.g. md5sum, sha1sum) in parallel,\n");
    printf("reducing repeated file I/O.\n");
    printf("This version of %s is compiled with support for the following checksum algorithms:\n", progname);
    for (i = 0; i < NUM_CHECKSUMS; i++)
    {
        struct checksum_info *info = checksum_information(i);
        printf("  %s (--%s)\n", info->name, info->prog);
    }
    exit (0);
}

int main(int argc, char **argv)
{
    int i = 0;
    int c = 0;
    int num_desired = 0;

    struct checksum_info *info= NULL;
    
    for (i = 0; i < NUM_CHECKSUMS; i++)
    {
        info = checksum_information(i);
        opts[i].name = strdup(info->prog);
	opts[i].has_arg = required_argument;
	opts[i].flag = NULL;
	opts[i].val = i;
    }

    opts[i].name = "help";
    opts[i].has_arg = no_argument;
    opts[i].flag = NULL;
    opts[i].val = i;
    i++;

    opts[i].name = NULL;
    opts[i].has_arg = 0;
    opts[i].flag = NULL;
    opts[i].val = 0;
    
    while (1)
    {
        int option_index = 0;
        c = getopt_long(argc, argv, "", opts, &option_index);
        if (c == -1 || c == '?')
        {
            break;
        }
	if (c == 4) {
            usage();
	}
	state[c].desired = 1;
	state[c].out_filename = strdup(optarg);
	state[c].out_stream = NULL;
	num_desired++;
	checksums_mask |= (1 << c);
    }

    if (0 == num_desired)
    {
        fprintf(stderr, "No checksum algorithms requested, bailing...\n");
	return 0;
    }

    if (optind >= argc)
    {
        /* nothing to do! */
        fprintf(stderr, "No files listed, nothing to do\n");
        return 0;
    }

    /* else - we have to work to do! */

    /* Open all the output files we want */
    for (i = 0; i < NUM_CHECKSUMS; i++)
    {
        if (state[i].desired)
	{
            state[i].out_stream = fopen(state[i].out_filename, "wb");
            if (NULL == state[i].out_stream)
	    {
                fprintf(stderr, "Can't open checksum file %s for writing, %s\n",
		        state[i].out_filename, strerror(errno));
	        return errno;
	    }
        }
    }

    /* Now work through the list of input files. */
    while (optind < argc)
    {
        checksum_context_t *ctx = checksum_init_context(checksums_mask, argv[optind]);
        if (NULL == ctx)
        {
            fprintf(stderr, "Failed to allocate memory for checksum context. ABORT\n");
            return ENOMEM;
        }

        queue.infile = fopen(argv[optind], "rb");
        if (queue.infile)
        {
            /* Main loop: read data, update checksums */
            queue.read_idx = 0;
            queue.put_idx = 0;
            queue.get_idx = 0;
            queue.num_full = 0;
            pthread_mutex_init(&queue.mutex, NULL);
	    pthread_cond_init(&queue.notfull, NULL);
	    pthread_cond_init(&queue.notempty, NULL);
            pthread_create(&queue.reader, NULL, reader, NULL);
            pthread_create(&queue.processor, NULL, processor, ctx);

            /* stuff happens in the threads */
            pthread_join (queue.reader, NULL);
            pthread_join (queue.processor, ctx);
            fclose(queue.infile);
            checksum_final(ctx);

            for (i = 0; i < NUM_CHECKSUMS; i++)
            {
                if (state[i].desired)
	        {
                    fprintf(state[i].out_stream, "%s  %s\n",
                            checksum_hex(ctx, i),
                            argv[optind]);
                    fflush(state[i].out_stream);
                }
            }
            checksum_free_context(ctx);
            optind++;
        }
    }

    /* Done with all the input files. Flush and close all the output
     * checksum files. */
    for (i = 0; i < NUM_CHECKSUMS; i++)
    {
        if (state[i].desired)
	{
            fflush(state[i].out_stream);
            fclose(state[i].out_stream);
        }
    }

    printf("bytes_read: %lld, processed: %lld\n", bytes_read, bytes_proc);
    printf("Time taken: %3f seconds reading, %3f seconds processing\n",
           ((double)total_read / 1000000), ((double)total_proc / 1000000));

    /* All done */
    return 0;
}
